Fasterize engine/1734 and Fasterize engine/1725

abarre opened the pull request:
This fixes two issues : 
- async="async" attribute were not taken into account
- defer script were not correctly prioritized

abarre committed:
[fasterize/FasterizeEngine#1734] Fix loading priority between scripts.

glejeune commented on lib/js_defer.js:
```diff
@@ -496,14 +501,20 @@ deferJsNs.DeferJs.prototype.cloneScriptNode = function(opt_script_elem) {
     var script = me.cloneScriptNode(script_elem);
     script.setAttribute('type', 'text/javascript');
-    var asyncScript = script.getAttribute('async') === 'true' || script.getAttribute('async') === '';
+    var asyncScript = script.getAttribute('async') !== null;
```
this meen that `async` can be anything, including `false`, `'false'`, `undefined`, `'undefined'`, `'no'`, `'sync'` ...

abarre replied:
According to https://developer.mozilla.org/fr/docs/Web/API/Element/getAttribute#Notes : 

> À peu près tous les navigateurs (Firefox, Internet Explorer, Opera, Safari, Konqueror et iCab, sans que cette liste soit exhaustive) renvoient null lorsque l'attribut demandé n'existe pas sur l'élément spécifié. La spécification DOM indique cependant que la valeur de retour correcte dans ce cas est en fait lachaîne vide , et certaines implémentation DOM auront ce comportement. Par conséquent, vous devriez utiliser element.hasAttribute() pour vérifier l'existence d'un attribut avant d'appeler getAttribute() s'il est possible que l'attribut n'existe pas sur l'élément spécifié.

I will have a test with `hasAttribute()`

glejeune commented on lib/js_defer.js:
```diff
@@ -1266,10 +1278,26 @@ deferJsNs.DeferJs.prototype.registerScriptTags = function() {
     if (script.getAttribute('type') == this.psaScriptType_) {
-      this.addNode(script);
+      // async attribute is priority over defer attribute
+      if (script.getAttribute('async') !== null) {
```
See my comment above.

glejeune commented on lib/js_defer.js:
```diff
@@ -1266,10 +1278,26 @@ deferJsNs.DeferJs.prototype.registerScriptTags = function() {
+      else {
+        // add before deferjs script
+        this.addNode(script, syncScript);
+        syncScript++;
```
If the code declare (in this order) a defer then a sync scripts, they will both be at position 0. Is that what we whant ?

abarre replied:
The logic should be to insert the sync script before the defer script in the queue. Thus, inserting the sync script at the position 0, will move the defered script at the position 1.
I think that the implementation is correct. I will try the case to verify.

abarre replied:
In this case, when the sync script will be set at the position 0, the deferred script will be set at the position 1.
It's tested with the test using the defer_order.html file.

abarre committed:
[fasterize/FasterizeEngine#1725] Fix support of async script

async attribute is a boolean attribute. Thus, setting async=« true »,
async=« false » or async=« anything » all mean the same thing.
They enable the async behavior.

ref : https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attributes
